doc/neps/nep-0043-extensible-ufuncs.rst

```diff
@@ -0,0 +1,631 @@
+
+   * The step has to perform no special work for non-parametric dtypes.
+   * For example, if a loop adds two strings, it is necessary to define the
+     correct output (and possibly input) dtypes.  ``S5 + S4 -> S9``, while
+     an ``upper`` function has the signature ``S5 -> S5``.
+
+4. Preparing the actual iteration. This step is largely handled by ``NpyIter`` (the iterator).
+
+   * Allocate all ouputs and potentially copies (or buffers).
+   * Finds the best iteration order, which includes information such as
+     a broadcasted stride always being 0.
+
+5. *Loop-specific setup* may include for example:
+
+   * Clearing of floating point exception flags (if necessary),
+   * Possibly allocating temporary working space,
+   * Setting (and potentially finding) the inner-loop function.  Finding
+     the inner-loop function could allow specialized implementations in the
+     future.
+   * Signal whether the inner-loop requires the Python API, or whether
+     the GIL may be released.
+
+6. Run the DType specific *inner-loop*
+
+   * The loop may require access to additional data, such as dtypes or
+     additional data set in the previous step.
+
+7. *Loop-specific Teardown* may be necessary to undo any setup done in step 5
+   such as checking for floating point errors.
+
+This NEP proposes a new registration approach for step 2 by creating a private
+``UfuncImpl`` structure which will be filled using an extensible API,
+and which may be exposed as a Python object later.
+This shall allow users to define custom behaviour for steps 3, 5, and 7,
+while extending the inner-loop function (step 6) accordingly.
+
+The following sections go into more details, and are seperated into the
+two main topics of *promotion and dispatching* and the further C-API
+provided to the user for the ufunc execution.
+
+
+UFuncImpl Specifications
+""""""""""""""""""""""""
+
+These specifications provide a minimal initial API, which shall be expanded
+in the future, for example to allow specialized inner-loops.
+
+Briefly, NumPy currently relies fully on strided inner-loops and, this
+will be the only allowed method of defining a ufunc initially.
+With additional setup and teardown functionality, as well as the
+``adapt_descriptors`` function mirroring the same functionality as required
+for casting (see also NEP 42 ``CastingImpl``).
+This gives the following function definitions:
+
+* Similar to casting, also ufuncs may need to find the correct output DType
+  or indicate that a loop is only capable of handling certain instances of
+  the involved DTypes (e.g. only native byte order).  This is handled by
+  an ``resolve_descriptors`` function (similar to ``adjust_descriptors``
+  of ``CastingImpl``)::
+
+      int resolve_descriptors(
+              PyUFuncObject *ufunc, PyArray_DTypeMeta *DTypes[nargs],
+              PyArray_Descr *in_dtypes[nargs], PyArray_Descr *out_dtypes[nargs]);
+
+  setting the desired output descriptors, and possibly modifying the input
```

Comment:
@seberg - the "and possibly modifying the input" here was what made me think that changes were made in-place. Maybe rephrase this?